File: Java/Graphs/Detect cycle in undirected graph.java
import java.util.*;

public class Solution {
    private boolean dfs(int node, Queue<int[]> Q, int[] visited, ArrayList<ArrayList<Integer>> adjLs) {
        Q.offer(new int[]{node, -1});
        visited[node] = 1;

        while (!Q.isEmpty()) {
            int[] pair = Q.poll();
            int currentNode = pair[0];
            int parent = pair[1];

            for (int i : adjLs.get(currentNode)) {
                if (visited[i] == 0) {
                    Q.offer(new int[]{i, currentNode});
                    visited[i] = 1;
                } else if (parent != i) {
                    return true;
                }
            }
        }
        return false;
    }

    public static String cycleDetection(int[][] edges, int n, int m) {
        int[] visited = new int[n + 1];
        ArrayList<ArrayList<Integer>> adjLs = new ArrayList<>();

        for (int i = 0; i <= n; i++) {
            adjLs.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adjLs.get(u).add(v);
            adjLs.get(v).add(u);
        }

        Solution solution = new Solution();
        Queue<int[]> Q = new LinkedList<>();
        for (int i = 1; i <= n; i++) {
            if (visited[i] == 0) {
                if (solution.dfs(i, Q, visited, adjLs)) {
                    return "Yes";
                }
            }
        }
        return "No";
    }
}

File: Python /Graphs/Detect cycle in undirected graph.py
from collections import deque
def cycleDetection(edges, n, m):
    def isCycle(node):
        Q.append((node, -1))
        visited[node] = 1

        while Q:
            node, parent = Q.popleft()
            for i in adjLs[node]:
                if not visited[i]:
                    Q.append((i, node))
                    visited[i] = 1

                elif parent != i:
                    return True
        return False

    Q = deque()
    visited = [0]*(n+1)
    
    adjLs =[[]for _ in range(n+1)]
    for u, v in edges:
        adjLs[u].append(v)
        adjLs[v].append(u)

    for i in range(1, n+1):
        if not visited[i]:
            if isCycle(i):
                return "Yes"
    return "No"
